learning-another-language: content (#7)
Ada-Developers-Academy/core-unit-3: learning-another-language/fizzbuzz.md

# FizzBuzz

## Introduction

When learning another language, implementing FizzBuzz is a great way to jump into it. This program allows us to explore how to:

- use variables
- use conditionals
- use equality-checking and comparison
- define functions
- use parameters
- use `return`
- call functions

> The requirements of the FizzBuzz program is as follows:

Reviewer: numerosity
```suggestion
The requirements of the FizzBuzz program are as follows:
```

Create a function named `fizzBuzz`. It takes in one parameter, `num`.

- If `num` is a multiple of 3, then return `'Fizz'`.
- If `num` is a multiple of 5, then return `'Buzz'`.
- If `num` is a multiple of 3 _and_ 5, then return `FizzBuzz`.
- If `num` is not a multiple of 3 or 5, then return the number itself.

## Implementation

[To implement FizzBuzz, we can write](https://replit.com/@adacore/FizzBuzz):

<!-- prettier-ignore-start -->
```js
const fizzBuzz = function(num) {

>   if (num % 15 == 0) {
>     return 'FizzBuzz';
>   } else if (num % 3 == 0) {
>     return 'Fizz';
>   } else if (num % 5 == 0) {

Reviewer: prefer ===
```suggestion
  if (num % 15 === 0) {
    return 'FizzBuzz';
  } else if (num % 3 === 0) {
    return 'Fizz';
  } else if (num % 5 === 0) {
```

  } else {
    return num;
  }
}

console.log(fizzBuzz(1));
console.log(fizzBuzz(3));
console.log(fizzBuzz(5));
console.log(fizzBuzz(15));
```
<!-- prettier-ignore-end -->

[When we run our code](https://replit.com/@adacore/FizzBuzz), we should see this output:

```
1
Fizz
Buzz
FizzBuzz
```

### Variable Assignment and Function Definition

Let's dive into this fragment:

```js
const fizzBuzz = function(num) {
```

| <div style="min-width:200px;"> Piece of Code </div> | Notes                                                                                                                                   |
| --------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------- |
| `const fizzBuzz`                                    | This _declares_ a variable named `fizzBuzz`. This specific type of variable is a `const` variable. The variable's name is in camelCase. |
| `=`                                                 | This is the assignment operator. This will assign whatever value is on the right-side to the variable on the left side, `fizzBuzz`.     |

> | `function`                                          | The keyword to begin the definition of a function                                                                                       |

Reviewer: here, this is a function expression. I think we can add that small amount of precision without it being confusing
```suggestion
| `function`                                          | The keyword to begin the definition of a function expression                                                                                      |
```

| `(num)`                                             | This section defines the function parameters. This function has one parameter, named `num`.                                             |
| `{`                                                 | This opening curly brace (`{`) begins the function body.                                                                                |

## Conditionals, Comparisons, and Returns

Now, let's dive into this fragment of code inside the function body:

<!-- prettier-ignore-start -->
```js

> if (num % 15 == 0) {
>   return 'FizzBuzz';
> } else if (num % 3 == 0) {

Reviewer: prefer ===
```suggestion
if (num % 15 === 0) {
  return 'FizzBuzz';
} else if (num % 3 === 0) {
```

  return 'Fizz';
}
```
<!-- prettier-ignore-end -->

| <div style="min-width:200px;"> Piece of Code </div> | Notes                                                                                                                                                                 |
| --------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `if`                                                | The keyword to begin an if-statement.                                                                                                                                 |
| `(...)`                                             | We put the conditional expression in parentheses `()`. If the expression evaluates to `true`, then the if-statement will execute the block of code.                   |

> | `num % 15 == 0`                                     | The conditional expression. It reads the parameter `num` and uses the mod operator `%`. It checks if `num % 15` is equal to `0` using equality comparison, with `==`. |

Reviewer: reminder about how 15 ties back to the requirement
```suggestion
| `num % 15 == 0`                                     | The conditional expression. It reads the parameter `num` and uses the mod operator `%`. It checks if `num % 15` is equal to `0` using equality comparison, with `==`. This checks whether a number is divisible by 15, or in other words, whether a number is divisible by both 3 and 5 (since 3 * 5 = 15). |
```

> | `{`                                                 | We begin the block of code to execute with an open curly brace `{`. **Everything inside the code block is indented one level**.                                       |

Reviewer: yes, we should indent. but callout that this is only for readability (explicitly chose not to mention python by name).
```suggestion
| `{`                                                 | We begin the block of code to execute with an open curly brace `{`. **Everything inside the code block is indented one level**. In JavaScript, the indentation has no semantic meaning, but it still helps us understand the code intent.                                      |
```

| `return 'FizzBuzz';`                                | A return statement. This returns the string literal `'FizzBuzz'`. The return statement ends in a semicolon `;`.                                                       |
| `}`                                                 | The block of code to execute for the if-statement ends with a closing curly brace `}`. **The end of this code block is indented to match the opening `if`**.          |
| `else if`                                           | This begins an else-if-statement. If the first conditional didn't return `true`, we will check if this next conditional will.                                         |
| `(...)`                                             | Identical to before, this holds the conditional expression to check.                                                                                                  |

> | `num % 3 == 0`                                      | This conditional expression checks if `num` is divisible by `3`, as according to the problem statement.                                                               |

Reviewer: ```suggestion
| `num % 3 == 0`                                      | This conditional expression checks if `num` is divisible by `3`, as required by the problem statement.                                                               |
```

It's good to note that the function ended with a closing curly brace (`}`).

The indentation of this curly brace matches the indentation-level of the line that _began_ the function definition.

### Invoking More Functions

Now, let's dive into the series of print statements after the function.

<!-- prettier-ignore-start -->
```js
console.log(fizzBuzz(1));
console.log(fizzBuzz(3));
```
<!-- prettier-ignore-end -->

| <div style="min-width:200px;"> Piece of Code </div> | Notes                                                                                                                                                                                                     |
| --------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `console.log(...)`                                  | This invokes the function that prints output to the console. More specifically, it invokes the `log` function that is part of the `console` object. _It prints the value that is the passed-in argument_. |

> | `fizzBuzz(1)`                                       | This invokes the function `fizzBuzz` with the argument `1`. We can expect it to return the number `1`.                                                                                                    |

Reviewer: we didn't just pull this out of thin air
```suggestion
| `fizzBuzz(1)`                                       | This invokes the function `fizzBuzz` with the argument `1`. From the problem statement, we can expect it to return the number `1`.                                                                                                    |
```

| `console.log(...);`                                 | Again, this function call will print out whatever value is passed into it.                                                                                                                                |
| `fizzBuzz(3)`                                       | This invokes the function `fizzBuzz`, passing in `3`. Its eventual return value is `'Fizz'`.                                                                                                              |

With this section, we should feel a little more confident reading when functions are being invoked.

## Observe and Apply Patterns

We saw how we printed the result of calling `fizzBuzz` twice. We can apply those learnings to the remaining lines:

```js
console.log(fizzBuzz(5));
console.log(fizzBuzz(15));
```

## Check for Understanding

<!-- Question 1 -->
<!-- prettier-ignore-start -->
### !challenge
* type: multiple-choice
* id: a06614f2
* title: FizzBuzz
##### !question

Consider the FizzBuzz implementation again:

```js
const fizzBuzz = function(num) {

>   if (num % 15 == 0) {
>     return 'FizzBuzz';
>   } else if (num % 3 == 0) {
>     return 'Fizz';
>   } else if (num % 5 == 0) {

Reviewer: prefer ===
```suggestion
  if (num % 15 === 0) {
    return 'FizzBuzz';
  } else if (num % 3 === 0) {
    return 'Fizz';
  } else if (num % 5 === 0) {
```
